BEP18.md

```diff
@@ -0,0 +1,123 @@
+
+We propose this BEP to enhance full node user experience (and ease their pain) on using state sync because of following implementation limitations.
+
+1. Users complain most about state syncing testnet is very slow and usually stuck on some requests. <br> <br> In this enhancement, we want data to respond more evenly across peers so that syncing can continuously make progress and the overall syncing time can reduce from 30 - 45 min to around 5 min.
+
+2. Interruption during state sync (node process get killed because of reboot computer or user impatience) would make already synced data in vain (because the current full node doesn't persist synced part on disk). Worsely it mistakenly writes a lock file prevents user state sync again. <br> <br> In this enhancement, we want support break-resume downloading and keep the consistent status for arbitrarily restart.
+
+## 5.  Specification
+
+State sync will download **manifest** and **snapshot chunks** from other peers.
+
+### 5.1 Take snapshot
+
+There are two ways to take snapshots from a fullnode: automatically or manually. Snapshots will be put under `$HOME/data/snapshot/<height>`. All types involved in the snapshot are encoded by go-amino and compressed by snappy. More details will be explained later.
+
+1. To make fullnode automatically take snapshots, just make sure `state_sync_reactor` in `$HOME/config.toml` is set to true.
+2. To manually take snapshots, stop the node if it is running, then run `./bnbchaind snapshot --home <home> --height <height>`.
+
+If the snapshot taking procedure is interrupted, the node will be still in good status, but it cannot provide the interrupted height for other peers.
+
+Note: automatically take snapshot would keep occupying disk space. Fullnode would not delete them automatically, so the user should periodically delete unneeded snapshots manually if they want to save disk space.
+
+### 5.2 Sync snapshot
+
+Syncing snapshot is designed to be only run once during full node first start. To enable state sync from others, `state_sync_reactor` should be true and `state_sync_height` should be set to non-negative (default `-1` means disable syncing from others).
+
+If a user wants to sync from (majority) peers' latest sync-able height, they should set `state_sync_height` to 0.
+
+Stop and restart fullnode during state sync is allowed. The next time full node is started, it will resume by loading Manifest and downloaded chunks then download missing chunks.
+
+Once state sync is successful, a `STATESYNC.LOCK` file will be created under `$HOME/data` to prevent state sync next time.
+
+### 5.3 Manifest format
+
+Manifest serves as a summary of snapshot chunks to be synced. It also maintains the order and types of snapshot chunks. Fullnode firstly asks peer's for the manifest file at the beginning of state sync and will trust majority peers with the same manifest.
+
+SHA256 hash sum of each chunk synced will be checked against the hash declared within the manifest file.
```

Comment:
Does it make sense to hash the manifest file itself and give the user ability to provide that hash when it wants to sync to a specific height?

Reply:
> Does it make sense to hash the manifest file itself and give the user the ability to provide that hash when it wants to sync to a specific height?

actually, the user can set `state_sync_height` in `config.toml`to the snapshot height other peers have. But the height number is not fixed for every day, we are considering how to let node knowns what snapshot heights I can provide.

The height is much simpler than the hash of manifest right?

Reply:
Yeah, but height + hash might give the user additional insurance (like when the nodes the user trust publish height + height of the snapshot and the user uses them to sync the state)